ddtrace/contrib/internal/google_generativeai/_utils.py

```diff
@@ -0,0 +1,219 @@
+import sys
+
+import wrapt
+
+from ddtrace.internal.utils import get_argument_value
+
+
+class BaseTracedGenerateContentResponse(wrapt.ObjectProxy):
+    """Base wrapper class for GenerateContentResponse objects for tracing streamed responses."""
+
+    def __init__(self, wrapped, instance, integration, span, args, kwargs):
+        super().__init__(wrapped)
+        self._model_instance = instance
+        self._dd_integration = integration
+        self._dd_span = span
+        self._args = args
+        self._kwargs = kwargs
+
+
+class TracedGenerateContentResponse(BaseTracedGenerateContentResponse):
+    def __iter__(self):
+        try:
+            for chunk in self.__wrapped__.__iter__():
+                yield chunk
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            raise
+        else:
+            tag_response(self._dd_span, self.__wrapped__, self._dd_integration, self._model_instance)
+        finally:
+            self._dd_span.finish()
+
+
+class TracedAsyncGenerateContentResponse(BaseTracedGenerateContentResponse):
+    async def __aiter__(self):
+        try:
+            async for chunk in self.__wrapped__.__aiter__():
+                yield chunk
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            raise
+        else:
+            tag_response(self._dd_span, self.__wrapped__, self._dd_integration, self._model_instance)
+        finally:
+            self._dd_span.finish()
+
+
+def _extract_model_name(instance):
+    """Extract the model name from the instance."""
+    model_name = getattr(instance, "model_name", "")
+    if not model_name or not isinstance(model_name, str):
+        return ""
+    if "/" in model_name:
```

Comment:
Is this processing for the model name necessary? I assume users would just want the full model name

Reply:
I do agree that processing is good to be able to target cases based on normalized model names. However we lose information about if it's a proprietary or what provider it has (ex: tfhub/EfficientNetV2 vs Proprietary/EfficientNetV2). So I'm suggesting if we can also extract the model provider along with the name ?

Reply:
Will add a comment in the code to clarify why we're doing this processing!